### PR TITLE
improve and load plugin command stubs when required

### LINKED ISSUE #3621: Performance regression on every command invocation
```
cbec75e2f3f5afd3334db693829a6b150612076c is the first bad commit
commit cbec75e2f3f5afd3334db693829a6b150612076c
Author: Nicolas De Loof <nicolas.deloof@gmail.com>
Date:   Wed Mar 30 15:27:25 2022 +0200

    Adopt Cobra completion v2 to support completion by CLI plugins

    Signed-off-by: Nicolas De Loof <nicolas.deloof@gmail.com>

 cli-plugins/manager/cobra.go        | 48 +++++++++++++++---
 cli-plugins/plugin/plugin.go        |  5 ++
...
```

After https://github.com/docker/cli/pull/3429 every command invocation takes ~200ms longer. Even when command just redirects to another process. 

Before:

```
 # time /tmp/docker buildx version
github.com/docker/buildx v0.8.0-92-g908ce2d2.m 908ce2d20618a069ddd06fc4f044505899932fe2.m
/tmp/docker buildx version  0.06s user 0.04s system 98% cpu 0.098 total
 # time /tmp/docker buildx version
github.com/docker/buildx v0.8.0-92-g908ce2d2.m 908ce2d20618a069ddd06fc4f044505899932fe2.m
/tmp/docker buildx version  0.06s user 0.03s system 99% cpu 0.095 total
 # time /tmp/docker buildx version
github.com/docker/buildx v0.8.0-92-g908ce2d2.m 908ce2d20618a069ddd06fc4f044505899932fe2.m
/tmp/docker buildx version  0.06s user 0.03s system 94% cpu 0.096 total
 # time /tmp/docker --version
Docker version unknown-version, build unknown-commit
/tmp/docker --version  0.02s user 0.01s system 71% cpu 0.038 total
 # time /tmp/docker --version
Docker version unknown-version, build unknown-commit
/tmp/docker --version  0.02s user 0.01s system 70% cpu 0.036 total
```

After:
```
github.com/docker/buildx v0.8.0-92-g908ce2d2.m 908ce2d20618a069ddd06fc4f044505899932fe2.m
/tmp/docker buildx version  0.19s user 0.10s system 89% cpu 0.322 total
 # time /tmp/docker buildx version
github.com/docker/buildx v0.8.0-92-g908ce2d2.m 908ce2d20618a069ddd06fc4f044505899932fe2.m
/tmp/docker buildx version  0.19s user 0.10s system 90% cpu 0.325 total
 # time /tmp/docker buildx version
github.com/docker/buildx v0.8.0-92-g908ce2d2.m 908ce2d20618a069ddd06fc4f044505899932fe2.m
/tmp/docker buildx version  0.18s user 0.10s system 90% cpu 0.312 total
 # time /tmp/docker --version
Docker version unknown-version, build unknown-commit
/tmp/docker --version  0.16s user 0.09s system 89% cpu 0.280 total
 # time /tmp/docker --version
Docker version unknown-version, build unknown-commit
/tmp/docker --version  0.16s user 0.09s system 89% cpu 0.272 total
```

@ndeloof

### PR DESCRIPTION
**Pull request by crazy-max**
fixes #3621 
related to https://github.com/docker/cli/pull/3429
closes #4119
closes #4120

**- What I did**

We are currently loading plugin commands stubs for every command invocation to add support for Cobra v2 completion. This causes a significant performance hit if there are a lot of plugins in the user space (7 atm in Docker Desktop):

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc., v0.10.3)
  compose: Docker Compose (Docker Inc., v2.15.1)
  dev: Docker Dev Environments (Docker Inc., v0.1.0)
  extension: Manages Docker extensions (Docker Inc., v0.2.18)
  sbom: View the packaged-based Software Bill Of Materials (SBOM) for an image (Anchore Inc., 0.6.0)
  scan: Docker Scan (Docker Inc., v0.25.0)
  scout: Command line tool for Docker Scout (Docker Inc., v0.6.0)
```

`docker --version` takes ~93ms currently:

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `docker-20.10.12` | 19.2 ± 1.3 | 18.4 | 21.4 | 2.04 ± 0.47 |
| `docker-20.10.17` | 19.5 ± 0.8 | 18.5 | 20.5 | 2.07 ± 0.47 |
| `docker-20.10.23` | 15.3 ± 2.4 | 12.6 | 19.0 | 1.63 ± 0.44 |
| `docker-23.0.1` | 93.9 ± 1.8 | 91.0 | 95.8 | 9.96 ± 2.22 |
| `docker-dev-pr-3419-a4b6fe1` | 17.5 ± 1.0 | 16.8 | 19.1 | 1.85 ± 0.42 |
| `docker-dev-pr-3429-a09e61a` | 93.7 ± 0.8 | 92.5 | 94.5 | 9.94 ± 2.21 |

Instead of removing completion for plugins to fix the regression as suggested in #4119, we can slightly improve plugins discovery by spawning a goroutine for each iteration in the loop when listing plugins:

`docker --version` now takes ~38ms

But there is still a significant performance hit compared to `20.10.23`. With second commit we are now loading plugins only if cobra completion arg request is found and also make sure plugin command stubs are loaded once.

With this change `docker --version` now takes ~9ms:

| Command | Mean [ms] | Min [ms] | Max [ms] | Relative |
|:---|---:|---:|---:|---:|
| `docker-20.10.12` | 19.2 ± 1.3 | 18.4 | 21.4 | 2.04 ± 0.47 |
| `docker-20.10.17` | 19.5 ± 0.8 | 18.5 | 20.5 | 2.07 ± 0.47 |
| `docker-20.10.23` | 15.3 ± 2.4 | 12.6 | 19.0 | 1.63 ± 0.44 |
| `docker-23.0.1` | 93.9 ± 1.8 | 91.0 | 95.8 | 9.96 ± 2.22 |
| `docker-dev-pr-3419-a4b6fe1` | 17.5 ± 1.0 | 16.8 | 19.1 | 1.85 ± 0.42 |
| `docker-dev-pr-3429-a09e61a` | 93.7 ± 0.8 | 92.5 | 94.5 | 9.94 ± 2.21 |
| `docker-dev-fix-perf-reg-2` | 9.4 ± 2.1 | 7.6 | 12.6 | 1.00 |

We should also look if every plugin we currently ship doesn't introduce performance regressions when invoked through the plugin manager in https://github.com/docker/cli/blob/f5d698a331f520e6fa1583911e5dedff428693f3/cli-plugins/manager/candidate.go#L22 to load metadata:

```
/usr/local/lib/docker/cli-plugins/docker-buildx docker-cli-plugin-metadata
```

Full benchmark can be seen at https://github.com/crazy-max/docker-cli-bench/blob/main/bench.md

**- How to verify it**

```
$ make -f docker.Makefile binary
```

Check completion still working for commands and plugins per #3429:

```
$ ./build/docker __completeNoDesc "b"
build
builder
buildx
:4
Completion ended with directive: ShellCompDirectiveNoFileComp

$ ./build/docker __completeNoDesc "c"
checkpoint
commit
compose
config
container
context
cp
create
:4
Completion ended with directive: ShellCompDirectiveNoFileComp

$ ./build/docker __complete --context "de"
./build/docker __complete --context "de"
desktop-linux
default
:4
Completion ended with directive: ShellCompDirectiveNoFileComp
```

Check a plugin help output:

```
$ ./build/docker buildx bake --help

Usage:  docker buildx bake [OPTIONS] [TARGET...]

Build from a file

Aliases:
  docker buildx bake, docker buildx f

Options:
      --builder string         Override the configured builder instance
  -f, --file stringArray       Build definition file
      --load                   Shorthand for "--set=*.output=type=docker"
      --metadata-file string   Write build result metadata to the file
      --no-cache               Do not use cache when building the image
      --print                  Print the options without building
      --progress string        Set type of progress output ("auto", "plain", "tty"). Use plain to show container output (default "auto")
      --provenance string      Shorthand for "--set=*.attest=type=provenance"
      --pull                   Always attempt to pull all referenced images
      --push                   Shorthand for "--set=*.output=type=registry"
      --sbom string            Shorthand for "--set=*.attest=type=sbom"
      --set stringArray        Override target value (e.g., "targetpattern.key=value")
```

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by bsousaa** [comment]
cc @laurazard

File: cmd/docker/docker.go [hunk 1]
```diff
@@ -227,9 +234,14 @@ func runDocker(dockerCli *command.DockerCli) error {
...
 	}
 
-	err = pluginmanager.AddPluginCommandStubs(dockerCli, cmd)
-	if err != nil {
-		return err
+	if cli.HasCompletionArg(args) {
+		// We add plugin command stubs early only for completion. We don't
+		// want to add them for normal command execution as it would cause
+		// a significant performance hit.
+		err = pluginmanager.AddPluginCommandStubs(dockerCli, cmd)
```
Comment (tonistiigi) [inline cmd/docker/docker.go:241]: Even if there is a completion arg, why do we need to list all the plugins? Most  likely the command is only for a specific plugin. Asking for follow-up, not a blocker.

Comment (crazy-max) [reply cmd/docker/docker.go:241]: Because we need to list completion values for plugins too so they have to be registered first.

Comment (tonistiigi) [reply cmd/docker/docker.go:241]: Only when the args is `docker`. Not when it is already `docker buildx`. This is not that important as only affects the completion path but ideally it shouldn't be that adding more plugins makes completion for another plugin slower.

Comment (crazy-max) [reply cmd/docker/docker.go:241]: Yes indeed, we could improve this.

**Review by tonistiigi — APPROVED** [review]

**Review by laurazard — APPROVED** [review]
Tested and changes LGTM

**Review by vvoland — APPROVED** [review]

**Comment by crazy-max** [comment]
Could we have cherry pick label for this?

**Comment by thaJeztah** [comment]
THANKS! Let's bring this one in 👍

**Pull request merged**